iceoryx_utils/include/iceoryx_utils/concurrent/resizeable_lockfree_queue.hpp

```diff
@@ -0,0 +1,198 @@
+    // 3) The final overload discards removed elements.
+
+    /// @brief      Set the capacity to some value.
+    /// @param[in]  newCapacity capacity to be set
+    /// @param[in]  removeHandler is a function taking an element which specifies
+    ///             what to do with removed elements should the need for removal arise.
+    /// @return     true if the capacity was successfully set, false otherwise
+    template <typename Function,
+              typename = typename std::enable_if<cxx::is_invocable<Function, ElementType>::value>::type>
+    bool setCapacity(uint64_t newCapacity, Function&& removeHandler) noexcept;
+
+    /// @brief Set the capacity to a new capacity between 0 and MaxCapacity, if the capacity is reduced
+    /// it may be necessary to remove the least recent elements.
+    /// @param[in] newCapacity new capacity to be set, if it is larger than MaxCapacity the call fails
+    /// @param[out] removedElements container were potentially removed elements can be stored.
+    /// @return true setting if the new capacity was successful, false otherwise (newCapacity > MaxCapacity)
+    /// @note threadsafe, lockfree but multiple concurrent calls may have no effect
+    template <typename ContainerType = iox::cxx::vector<ElementType, MaxCapacity>,
+              typename = typename std::enable_if<!cxx::is_invocable<ContainerType, ElementType>::value>::type>
+    bool setCapacity(uint64_t newCapacity, ContainerType& removedElements) noexcept;
+
+    /// @brief Set the capacity to a new capacity between 0 and MaxCapacity, if the capacity is reduced
+    /// it may be necessary to remove the least recent elements which are then discarded.
+    /// @param[in] newCapacity new capacity to be set, if it is larger than MaxCapacity the call fails
+    /// @return true setting if the new capacity was successful, false otherwise (newCapacity > MaxCapacity)
+    /// @note threadsafe, lockfree but multiple concurrent calls may have no effect
+    bool setCapacity(uint64_t newCapacity) noexcept;
+
+  private:
+    using BufferIndex = typename Base::BufferIndex;
+    std::atomic<uint64_t> m_capacity{MaxCapacity};
+
+    // we also sync m_capacity with this flag
+    std::atomic_flag m_resizeInProgress{false};
+
+    // Remark: The vector m_unusedIndices is protected by the atomic flag, but this also means dying during a resize
+    // will prevent further resizes (This is not a problem for the use case were only the dying receiver itself requests
+    // the resize.)
+    // I.e. resize is lockfree, but not in a useful and robust way as it assumes that a concurrent resize will
+    // always eventually complete (which is true when the application does not die and the relevant thread is scheduled
+    // eventually. The latter is the case for any OS and mandatory for a Realtime OS.
+
+    iox::cxx::vector<BufferIndex, MaxCapacity> m_unusedIndices;
+
+    /// @brief      Increase the capacity by some value.
```

Comment:
Yes, this was temporarily public and could still be. But I feel it is confusing to the user and hence decided against it.

I am unsure about it, I documented it because it may not be immediately obvious how it behaves for anyone reading it. Will remove the doxygen description then but leave some note or explanation.

Reply:
I leave it in for now, the methods are nontrivial in their semantics and used by setCapacity. This makes it easier to understand how setCapacity works I think.